content/blog/towards-continuous-performance-regression-testing.asciidoc

```diff
@@ -0,0 +1,594 @@
+simplifying the filtering and aggregation of recorded events before matching them against expected values.
+JfrUnit also persists the JFR recording of each test method,
+so you can examine it after a test failure,
+for instance using https://openjdk.java.net/projects/jmc/[JDK Mission Control].
+To learn more about JfrUnit and its capabilities, take a look at the project's https://github.com/gunnarmorling/jfrunit[README].
+The project is in an early proof-of-concept stage at the moment,
+so changes to its APIs and semantics are likely.
+
+Now that you've taken the JfrUnit quick tour, let's put that knowledge into practice.
+Our example project will be the https://github.com/gunnarmorling/jfr-custom-events[todo management] Quarkus application you may already be familiar with from my link:/blog/rest-api-monitoring-with-custom-jdk-flight-recorder-events/[earlier post] about custom JFR events.
+We're going to discuss multiple examples for using JfrUnit to identify potential performance regressions.
+
+== Spotting Increased Memory Allocation
+
+At first, let's explore how to identify increased memory allocation rates.
+Typically, it's mostly library and middleware authors who are interested in this,
+since it can make a huge difference for a library such as Hibernate ORM whether a method invoked many times on a hot code path allocates a few bytes more or less memory.
+Allocating less memory means less work for the garbage collector,
+which in turn means those precious CPU cores of your machine can spend more cycles processing your actual business logic.
+
+But also for an actual application itself it can be beneficial to keep an eye on -- and systematically track -- object allocations,
+as regressions there lead to increased GC pressure,
+and in turn eventually to higher latencies and reduced throughput.
+
+The key for tracking object allocations with JFR are the `jdk.ObjectAllocationInNewTLAB` and `jdk.ObjectAllocationOutsideTLAB` events,
+which are emitted when
+
+* an object allocation triggered the creation of a new thread-local allocation buffer (TLAB) 
+* an object got allocated outside of the threads TLAB
+
+[NOTE]
+.Thread-local allocation buffers (TLAB)
+====
+When creating new object instances on the heap,
+this primarily happens via _thread-local allocation buffers_.
+A TLAB is a pre-allocated memory block that's exclusively used by a single thread.
+Creating new objects within a TLAB can happen without costly synchronization with other threads.
+Once a thread's current TLAB capacity is about to be exceeded by a new object allocation,
+a new TLAB will be allocated for that thread.
+In addition, large objects can also be directly allocated, outside of the TLAB.
+
+To learn more about TLAB allocation, refer to  https://shipilev.net/jvm/anatomy-quarks/4-tlab-allocation/[part #4] of Aleksey Shipilёv's "JVM Anatomy Quark" blog series.
+====
+
+Note these events don't allow for tracking of each individual object allocation,
```

Comment:
Are you sure? I do remember chasing TLAB allocation in more detail, but indeed this was in the older JFR and I never really understood why this metric was now. Maybe that was a mistake and it was removed later?

Reply:
cc/ @stalep

Reply:
Yes, pretty sure. The "inside TLAB" event gives you the new TLAB size and the size of the object which caused this (hence the notion of sampling). The "outside TLAB" gives you the object size, perhaps that's what you have in mind?

Reply:
ah yes, I simply misunderstood what you meant. _sampling_ actually is an option. I misread this as suggesting one couldn't get any clue about what was getting into TLAB.